### ISSUE TITLE
New Repo: knative/hack

### ISSUE THREAD

**Issue by n3wscott** [knative/community#322]
## Repo information

Org: knative

Repo: hack

Purpose (Description): The shared scripts for Knative.

We are looking at ways to reduce the transient dependencies of Knative from test-infra. Hack will be the base lib to share common scripts for build and code gen, as well as running tests, but without the tools and config that exist in the current test-infra repo.

test-infra will dep on `hack` and drop the scripts dir.

Sponsoring WG: Productivity 

## Actions to fulfill

This area is used to track the [repo creation process](/mechanics/CREATING-A-SANDBOX-REPO.md).
The _requestor_ and _sponsoring WG lead_ should perform the steps listed below and cross out the checkmarks when done.
The TOC is involved only in the **TOC Gate** steps.

- [x] Add this issue to the [TOC project board](https://github.com/orgs/knative/projects/9) for review.

_You may not be able to use the Projects quick menu on this page. In that case, go to the project board and use the **Add cards** interface._

- [x] Send a PR adding entries for this repo in [`/peribolos/knative-sandbox.yaml`] (https://github.com/knative/community/blob/master/peribolos/knative-sandbox.yaml). Please mind the alphabetical order when adding to a list. - https://github.com/knative/community/pull/323
  - [x] Add the repository and a description.
  - [x] Grant `Knative Admin` the `admin` privilege.
  - [x] Grant the sponsoring WG the `write` privilege.

**TOC Gate**: _Once the TOC has approved the above, it will merge and Peribolos will create an empty repository._

- [x] (golang) Send a PR to add aliases for `knative.dev/$REPONAME` import paths ([sample](https://github.com/knative/website/pull/187)).

- [x] Have a lead from the sponsoring WG bootstrap the Git repository by pushing an
  appropriate "template" repository ([basic](https://github.com/knative-sandbox/wg-repository),
  [sample-controller](https://github.com/knative-sandbox/sample-controller),
  [sample-source](https://github.com/knative-sandbox/sample-source)) to the new repository as
  a git remote.  For example:

  ```shell
    git clone https://github.com/knative-sandbox/sample-controller.git
    cd sample-controller
    git remote add newrepo https://github.com/knative-sandbox/$REPONAME.git
    git push newrepo master
  ```

- [x] Set up [test-infra](https://github.com/knative/test-infra) following the
  docs linked at the beginning.

- [x] Create a sample PR to verify Prow (e.g. edit the boilerplate README)

**TOC Gate** : _Once Prow has been verified_

- [x] Have a member of the TOC enable branch protections so that `tide` is a required presubmit check.

**Comment by n3wscott** [comment]
Or we should just move these into knative.dev/pkg?

**Comment by chizhg** [comment]
I think it's better to make it a separate module in case there are repos that only depends on these scripts but not `pkg`.

**Comment by rhuss** [comment]
👍  to have the scripts separated from `test-infra` if this is sufficient to get rid on the `test-infra` dependencies in Knative projects.

When migrated, would it make sense to add a check to the github actions to ensure that this test-infra deps doesn't come back ? (happy to provide that check)

**Comment by n3wscott** [comment]
+1 I think the `hack` repo is the way to go, not just throw the scripts into `pkg`.

**Comment by slinkydeveloper** [comment]
+1000 for this new repo, since it can solve the issues we have with the circular dependencies https://github.com/knative/test-infra/issues/2414 https://github.com/kubernetes/test-infra/issues/19583

This repo should not contain any dependency nor to k8s upstream nor to knative to fix the circular dependency issue

**Comment by n3wscott** [comment]
> This repo should not contain any dependency nor to k8s upstream nor to knative to fix the circular dependency issue

We can make a big comment to prevent this. Likely little deps are ok, but not much outside the std lib for go. And not knative or k8s.

**Comment by slinkydeveloper** [comment]
@n3wscott i think we should experiment some tool to check circular dependencies on CI: if there is a circular dep with knative.dev, the build fails

**Comment by slinkydeveloper** [comment]
Some tool like that? https://github.com/divan/depscheck

**Comment by evankanderson** [comment]
I think we're agreed on creating this, but would like to have a name that's slightly more clue-ful as to what this does. (e.g. `ci-hack` or `ci-shell-pkg`).

**Comment by evankanderson** [comment]
(This will also need to be a go package to work with dep)

**Comment by n3wscott** [comment]
I am going to push back on your comment @evankanderson, `hack` is the right thing to call this, just like `pkg` is `pkg` and not `common-components-intended-for-reuse`. In go, the `hack` dir is where you look for scripts. That convention plus some documentation in the readme and there will be zero confusion. Plus, the normal way you integrate with this is via a clone of one of the sample repos, so you never even notice it.

`ci-hack` and `ci-shell-pkg` do not make sense because the the scripts don't only do `ci` things, they also do local testing and releases. `shell-pkg` implies there is code that you import and use in go, which is not correct. `scripts` is not correct because that is not how we talk about shell scripts in a repo in a golang project.

tl;dr: I think the best choice is still `hack`. Talking to several people, this is a common opinion.

**Comment by n3wscott** [comment]
go-imports: https://github.com/knative/website/pull/235

**Comment by n3wscott** [comment]
This is done as soon as the website PR is merged.

**Comment by mattmoor** [comment]
done
